File: number_guessing_ver4.py
import random
def secret_number(low, high):
    return random.randint(low,high)

def user_input(low, high):
    while True:
        try:
            x = int(input(f"Enter a number between {low} and {high} \n"))
            if x < low or x > high:
                print(f"Please enter a number between {low} and {high}")
                continue
            return x
        except ValueError:
            print("Enter a valid number")

def game_continuation():
    
    print(" If yes enter O \t If no enter X \n")
    while True:
        choice = input("Enter your choice: ").strip().lower()
        if choice in ("x", "o"):
            return choice
        print("Invalid choice! Please enter O or X only")

def show_difference(a, b):
    difference = abs(a - b)
    if difference <= 5:
        print("You are too close to the target")
    elif difference <= 10:
        print("You are close to the target")
    else:
        print("You are far from the target")

def play_again():
    print("Do you want to play again? ")
    return game_continuation()

def choose_level():
    while True:
        print("For easy level: enter 'e'\nFor medium level: enter 'm'\nFor Hard level: enter 'h'\n\n")
        level = input("Enter your choice: ").strip().lower()  
        if level in ("e", "m", "h"):
            return level
        else:
            print("Enter a valid choice")

def instructions(low, high, level):
    if level == "e":
        print(f"1. You have to guess a number between {low} to {high}\n2. You have 12 attempts left")
    elif level == "m":
        print(f"1. You have to guess a number between {low} to {high}\n2. You have only 10 attmepts left")
    else:
        print(f"1. You have to guess a number between {low} to {high}\n2. You have only 8 attempts left")

def main():
    print("----------Welcome to number guessing game---------")
    #print("----------Instructions:--------- \nFor Easy level:--\n1. You have to guess a number between 1 and 50\n2. You have only 12 attempts\nFor medium level:--\n1. You have to guess a number between 1 and 100\n2. You have only 10 attempts\nFor hard level:--\n1. You have to guess a number between 1 and 200\n2. You have only 8 attempts\n\nAfter every attempt you can choose whether to continue or exit\n")
    
    while True:
        level = choose_level()
        low = 1
        if level == "e":
            high = 50
            max_attempts = 12
        elif level == "m":
            high = 100
            max_attempts = 10
        else:
            high = 200
            max_attempts = 8
        instructions(low, high, level)
        print(f"You can choose to exit or continue the game after every attempt")
        target = secret_number(low, high)
        attempt = 0
        
        

        while True:
            user_num = user_input(low, high)
            
            attempt += 1
            if user_num == target:
                print(f"Congratulations you have cleared the game in {attempt} attempts")
                break
            elif user_num > target:
                print("The number you guessed was higher than the target")
            else:
                print("The number you guessed was lower than the target")

            if attempt == max_attempts:
                print("You could not clear the game.")
                print(f"The number you had to guess was {target}")
                break
            
            show_difference(user_num, target)
            print(f"You have attempted {attempt} times\n you have only {max_attempts - attempt} attempts left")
            print("Do you want to continue the game?")
            choice = game_continuation()
            if choice == "o":
                continue
            else:
                print(f"The number you had to guess was {target}\nThank you for playing ")
                return
          
        if play_again() == "o":
            continue
        else:
            break
    
if __name__ == "__main__":
    main()

File: number_guessing_ver4.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>

int user_input(int low, int high){
    int x;
    char input[50];
    char dummy;
    while(1){
        if(fgets(input, sizeof(input), stdin) == NULL){
            exit(1);
        }
        else{
            if(sscanf(input, "%d %1s", &x, &dummy) == 1){
                if (x > high || x < low){
                    printf("Please enter a number between %d and %d\n",low, high);
                }
                else{
                return x;
                }
            }
            else{
                printf("Please enter a valid number between %d and %d\n", low, high);
                }
            }
    }
}

char choose_level(){
    while(1){
        char input[50];
        char level;
        char dummy;
        printf("For easy level: enter 'e'\nFor medium level: enter 'm'\nFor Hard level: enter 'h'\n\n");
        if(fgets(input, sizeof(input), stdin) == NULL){
            exit(1);
        }
        else{
            if(sscanf(input, " %c %1s", &level, &dummy) == 1 && (level == 'e' || level =='E' || level == 'm' || level == 'M' || level == 'h' || level == 'H')){
                return level;
            }
            else{
                printf("You can only enter e, m, h\n");
            }
        }
    }
}

void instructions(char level, int low, int high){
    if (level == 'e' || level == 'E'){
        printf("1. You have to guess a number between %d and %d\n2. You have only 12 attempts\n", low,high);
    }
    else if (level == 'm' || level == 'M'){
        printf("1. You have to guess a number between %d and %d\n2. You have only 10 attempts\n", low, high);
    }
    else{
        printf("1. You have to guess a number between %d and %d\n2. You have only 8 attempts\n", low, high);
    }
}

char get_choice(){
    char input[50];
    char choice;
    char dummy;
    while(1){
        if(fgets(input, sizeof(input), stdin) == NULL){
            exit(1);
        }
        else{
            if(sscanf(input," %c %1s", &choice, &dummy) == 1 && (choice == 'x' || choice =='X' || choice == 'o' || choice == 'O')){
                return choice;
            }
            else{
                printf("Error! You can only enter X or O\n");
            }
        }
    }
}

void show_difference(int a, int b){
    int difference = abs(a - b);
    if(difference <=5){
        printf("You are too close\n");
    }
    else if(difference <= 10){
        printf("You are close\n");
    }
    else{
        printf("You are far from the target\n");
    }
}

char play_again(){
    printf("Do you want to play again\n Enter o if yes\t Enter x if no\n");
    return get_choice();
}

int main(){
        printf("Welcome to number guessing game\n");
        //printf("----Instructions----\n 1. You have to guess a number(integer) between 1 and 100\n 2. You have only 10 attempts\n 3. After every attempt you can choose to continue or exit the game\n");
        srand(time(NULL));
        while(1){


        char level;
        level = choose_level();
        int max_attempts;
        int low = 1;
        int high;
        if (level == 'e' || level == 'E'){
            max_attempts = 12;
            high = 50;
        }
        else if (level == 'm' || level == 'M'){
            max_attempts = 10;
            high = 100;
        }
        else{
            high = 200;
            max_attempts = 8;
        }
        instructions(level, low, high);
        printf("You can choose to continue or exit the game after every incorrect attempt\n");


        int guess, target;
        
        char choice;
        int attempt = 0;
        

        
        target = (rand() % high) + 1;
        
        while(1){
            attempt++;
            printf("Guess a number between 1 and %d\n", high);
            guess = user_input(low, high);
            if (guess == target){
                printf("Congratulations you cleared the game in %d attempts\n",attempt);
                break;
            }
            else if (guess > target){
                printf("The number was higher than the target\n");
            }
            else {
                printf("The number was lower than the target\n");
            }

            if(attempt == max_attempts){
                printf("You could not clear the game\n");
                printf("The number you had to guess was %d\n",target);
                break;
            }

            show_difference(target, guess);
            printf("You have attempted %d times\n",attempt);
            printf("Now you have only %d attempts\n",max_attempts - attempt);
            printf("Do you want to continue the game ?\n If Yes enter O\n If No enter X\n");
            choice = get_choice();
            if (choice == 'o' || choice =='O'){
                continue;
            }
            else{
                printf("The number you had to guess was %d\n Thank you for playing the game\n",target);
                return 0;
            }
        }
        char again = play_again();
        if (again == 'o' || again == 'O'){
            continue;
        }
        else{
            printf("Thank you for playing\n");
            break;
        }
    
    }
return 0;
}